lib/auth/auth.go

```diff
@@ -1670,6 +1670,27 @@ func (a *Server) SetAccessRequestState(ctx context.Context, params services.Acce
 	return trace.Wrap(err)
 }
 
+func (a *Server) GetAccessCapabilities(ctx context.Context, req services.AccessCapabilitiesRequest) (services.AccessCapabilities, error) {
+	var caps services.AccessCapabilities
+	if req.RequestableRoles {
+		// in order to determine which roles a user is allowed to request, we
+		// create a fake wildcard request and run validation with variable expansion
+		// turned on.
+		// TODO(fspmarshall): Refactor access request validation logic to allow
+		// calculating requestable roles more cleanly.
+		dummyReq, err := services.NewAccessRequest(req.User, services.Wildcard)
+		if err != nil {
+			return services.AccessCapabilities{}, trace.Wrap(err)
+		}
+		err = services.ValidateAccessRequest(a, dummyReq, services.ExpandRoles(true))
```

Comment:
TODO: suppress the error case where the user isn't allowed to request any roles.  This should just result in an empty `RequestableRoles` list instead.